llms/mlx_lm/tuner/dora.py

```diff
@@ -0,0 +1,119 @@
+import math
+from typing import Any
+
+import mlx.core as mx
+import mlx.nn as nn
+
+
+
+class DoRALinear(nn.Module):
+    @staticmethod
+    def from_linear(
+        linear: nn.Linear,
+        r: int = 8,
+        alpha: float = 16,
+        dropout: float = 0.0,
+        scale: float = 10.0,
+    ):
+        # TODO remove when input_dims and output_dims are attributes
+        # on linear and quantized linear
+        output_dims, input_dims = linear.weight.shape
+        if isinstance(linear, nn.QuantizedLinear):
+            input_dims *= 32 // linear.bits
+        dora_lin = DoRALinear(
+            linear=linear,
+            input_dims=input_dims,
+            output_dims=output_dims,
+            r=r,
+            alpha=alpha,
+            dropout=dropout,
+            scale=scale,
+        )
+        #dora_lin.linear = linear
+        return dora_lin
+    
+    
+    def to_linear(self, de_quantize: bool = False):
+        linear = self.linear
+        bias = "bias" in linear
+        weight = linear.weight
+        m=self.m
+
+        is_quantized = isinstance(linear, nn.QuantizedLinear)
+
+        # Use the same type as the linear weight if not quantized
+        dtype = weight.dtype
+
+        if is_quantized:
+            dtype = mx.float16
+            weight = mx.dequantize(
+                weight,
+                linear.scales,
+                linear.biases,
+                linear.group_size,
+                linear.bits,
+            )
+        output_dims, input_dims = weight.shape
+        fused_linear = nn.Linear(input_dims, output_dims, bias=bias)
+
+        lora_b = (self.scale * self.lora_b.T).astype(dtype)
+        lora_a = self.lora_a.T.astype(dtype)
+        fused_linear.weight = weight + lora_b @ lora_a
+
+        fused_norm=fused_linear.weight.square().sum(axis=0,keepdims=True).sqrt()
```

Comment:
`mx.linalg.norm`

Reply:
use `mx.linalg.norm` instead of `.square().sum(axis=1,keepdims=True).sqrt()`
```
        fused_linear.weight = weight + lora_b @ lora_a

        fused_norm=mx.linalg.norm(fused_linear.weight,axis=1)
```